### ISSUE TITLE
MagicDNS resolution fails when running inside docker 

### ISSUE THREAD

**Issue by MarcoRizk** [tailscale/tailscale#8535]
### What is the issue?

I am running tailscale inside docker, when i try to ping any of my machines by their name the ping fails giving ` Name or service not known`

The tailscale bugreport outputs the following

`user bugreport: BUG-eda763d395610760c93315317409b920013359249d5d1e1d77026ea8fe5181e7-20230705203058Z-b965faca85656897
user bugreport hostinfo: {"IPNVersion":"1.44.0-t5d008cd83-g6dac82fa0","OS":"linux","OSVersion":"5.19.0-46-generic","Container":true,"Distro":"debian","DistroVersion":"9.13","DistroCodeName":"stretch","Desktop":false,"Hostname":"1011b277d97b","Machine":"x86_64","GoArch":"amd64","GoArchVar":"v1","GoVersion":"go1.20.5-ts40dc4d8"}
user bugreport health: multiple errors:
        dns-os: getting OS base config is not supported
        dns: getting OS base config is not supported
user bugreport node info: nodeid="nodeid:1f366ea116046d" stableid="nnajoi6CNTRL" expiry="2024-01-01T20:11:12Z"
user bugreport public keys: machine="mkey:cc3f67d1746beaf192881ddc1b023a50e299f7f9f9643db1c4a85f6f2d845c20" node="nodekey:f7ca8b675dbe7d873216a4de1f1679dd28ba30464e7d3013c196797c749f3d02"
BUG-eda763d395610760c93315317409b920013359249d5d1e1d77026ea8fe5181e7-20230705203058Z-b965faca85656897`

### OS

Linux

### OS version

slim-stretch

### Bug report

BUG-eda763d395610760c93315317409b920013359249d5d1e1d77026ea8fe5181e7-20230705203058Z-b965faca85656897

**Comment by DentonGentry** [comment]
Indeed, this would be a problem:
```
2023-07-05 20:30:58.215276334 +0000 UTC: user bugreport health: multiple errors:
        dns-os: getting OS base config is not supported
        dns: getting OS base config is not supported
```

The tailnet does not have "Override local DNS" set, so it needs each node to retries the DNS servers to use from the platform but this container says it can't.

What is in /etc/resolv.conf in this container?

**Comment by MarcoRizk** [comment]
I added tailscale nameserver and search while running the container so resolv.conf has the following

`search DSL2750B tail5c505.ts.net
nameserver 192.168.1.1
nameserver 100.100.100.100
`
yet it still can't resolve the machine names. Any suggestions ?

**Comment by DentonGentry** [comment]
```
nameserver 192.168.1.1
nameserver 100.100.100.100
```

This shouldn't happen: if tailscaled [decides it needs to take over as the DNS server](https://tailscale.com/blog/sisyphean-dns-client-linux/) it should configure itself with `192.168.1.1` as an upsteam but not write it to /etc/resolv.conf.

Having both entries in the file means that it is up to the OS what happens.
- Some will send to all of the upstream servers and use which ever one responds first
- Some will send in order, with a timeout. Since 192.168.1.1 will probably respond with SERVFAIL for hostname.tailsc-scales.ts.net, this could be why MagicDNS names don't work.

**Comment by Bostrolicious** [comment]
This may not be relevant in this case, but I could not get MagicDNS to work inside a container until I disabled userspace networking with `TS_USERSPACE=false` in my Docker Compose file. I haven't seen that mentioned in any of the discussions I found about MagicDNS container problems, so hopefully it's helpful to someone.

**Comment by Tallisado** [comment]
`TS_USERSPACE=false` causes other issues it seems:
```
health("router"): error: setting up filter/ts-input: running [/sbin/iptables -t filter -N ts-input --wait]: exit status 3: iptables v1.8.9 (legacy): can't initialize iptables table `filter': iptables who? (do you need to insmod?)
```
The example given here also references the change mentioned, but throws a number of errors when run with identical config: https://github.com/tailscale-dev/docker-guide-code-examples/blob/main/02-oauth/compose.yaml

I'm on osx, using latest tailscale image. Might try a few versions back instead

**Comment by raggi** [comment]
> `TS_USERSPACE=false` causes other issues it seems:
> 
> ```
> health("router"): error: setting up filter/ts-input: running [/sbin/iptables -t filter -N ts-input --wait]: exit status 3: iptables v1.8.9 (legacy): can't initialize iptables table `filter': iptables who? (do you need to insmod?)
> ```
> 
> The example given here also references the change mentioned, but throws a number of errors when run with identical config: https://github.com/tailscale-dev/docker-guide-code-examples/blob/main/02-oauth/compose.yaml
> 
> I'm on osx, using latest tailscale image. Might try a few versions back instead

This issue is fixed by #11381.

**Comment by raggi** [comment]
Netstack mode (TS_USERSPACE=true) does not configure DNS resolution in the host environment because the DNS results would point to non-routable addresses. Outbound from the host connections in userspace mode are typically performed via SOCKS that will also handle name resolution. There was indeed a bug in tun mode for some of these cases, where the host system lacked IPv6 iptables netfilter support, which has been fixed as noted in the comment above.
